alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+    return false;
+}
+
+// expire_item() expires an item - i.e., deletes it as appropriate for
+// expiration - with CL=QUORUM and (FIXME!) in a way Alternator Streams
+// understands it is an expiration event - not a user-initiated deletion.
+static future<> expire_item(service::storage_proxy& proxy,
+                            const service::query_state& qs,
+                            const std::vector<bytes_opt>& row,
+                            const std::vector<unsigned>& pk_columns,
+                            const std::vector<unsigned>& ck_columns,
+                            schema_ptr schema,
+                            api::timestamp_type ts) {
+    // Prepare the row key to delete
+    std::vector<bytes> exploded_pk;
+    for (unsigned c : pk_columns) {
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
+            return make_ready_future<>();
+        }
+        exploded_pk.push_back(*row_c);
+    }
+    auto pk = partition_key::from_exploded(exploded_pk);
+    mutation m(schema, pk);
+    // If there's no clustering key, a tombstone should be created directly
+    // on a partition, not on a clustering row - otherwise it will look like
+    // an open-ended range tombstone, which will crash on KA/LA sstable format.
+    // See issue #6035
+    if (ck_columns.empty()) {
+        m.partition().apply(tombstone(ts, gc_clock::now()));
+    } else {
+        std::vector<bytes> exploded_ck;
+        for (unsigned c : ck_columns) {
+            const auto& row_c = row[c];
+            if (!row_c) {
+                // This shouldn't happen - all key columns must have values.
+                // But if it ever happens, let's just *not* expire the item.
+                return make_ready_future<>();
+            }
+            exploded_ck.push_back(*row_c);
+        }
+        auto ck = clustering_key::from_exploded(exploded_ck);
+        m.partition().clustered_row(*schema, ck).apply(tombstone(ts, gc_clock::now()));
+    }
+    return proxy.mutate(std::vector<mutation>{std::move(m)},
+        db::consistency_level::LOCAL_QUORUM,
+        executor::default_timeout(), // FIXME - which timeout?
+        qs.get_trace_state(), qs.get_permit());
+}
+
+
+future<> expiration_service::run() {
+    // FIXME: don't just tight-loop, think about timing, pace, and
+    // store position in durable storage, etc.
+    // FIXME: If one of the iterations below fails with an exception,
+    // this service will die forever. It's probably not a good idea!
+    // We should log the failure, but not stop the service.
+    for (;;) {
+        // FIXME: think about working on different tables in parallel.
+        // also need to notice when a new table is added, a table is
+        // deleted or when ttl is enabled or disabled for a table!
+        for (const auto& cf : _db.get_column_families()) {
```

Comment:
I'm reasonably sure that in case when no tables specify a TTL it's still reasonably safe to go through this whole loop without a single preemption point, but we might want to revise that later when we strive for multi-tenancy. E.g. if there could be a million tables in the system, this tight loop might cause a potential reactor stall, because I think that it only has preemption points on the path that actually expires a TTL. Adding an arbitrary preemption point is easy with coroutines (`co_await coroutine::maybe_yield()`), but doing so exceedingly often is also bad, because it adds overhead. To sum up, I think that no action is necessary for now, I'm just leaving this note here for other reviewers to consider, maybe somebody has a different opinion on this subject.

Also, I see that you consider adding parallelism to this loop, in which case the problem would probably disappear on its own anyway.

Reply:
I'll add a call to `co_await coroutine::maybe_yield()` inside the loop, just in case.